### PR TITLE
[Oxidize DAGCircuit] Implement `layer` methods and rust-native `op_nodes` for oxidiced `DAGCircuit`

### PR DESCRIPTION
**Pull request by raynelfss**
### Summary
The following commits aim to add `.layers`, `.multigraph_layers`, `.front_layer`, and `.serial_layers` to the oxidiced `DAGCircuit`. 

Also, rust native `op_nodes()` and `topilogical_op_nodes()` methods were added as well.

### Details and comments
Work in progress

### CODE REVIEW

**Comment by jlapeyre** [comment]
@raynelfss  I merged the current branch `oxidize-dag` into this branch (locally) and pushed it to 
https://github.com/jlapeyre/qiskit-core/tree/oxidize-dag-layers-merge-base

This PR #16 is blocking some debugging of `oxidize-dag` because the `todo!` in `fn layers` is hit so frequently. 
When I test the branch above, I find that far more tests run that with the current `oxidize-dag` https://github.com/kevinhartman/qiskit/tree/oxidize-dag

If you can use the branch referenced above it would unblock work and probably making merging less painful later. If I make a PR to this branch, tons of unrealted commits will appear.

Maybe if you pull https://github.com/jlapeyre/qiskit-core/tree/oxidize-dag-layers-merge-base and push it to the branch o this PR the authorships will be preserved.

I don't have any idea if it's feasible with your plans, but you might consider splitting this into two PRs so that `layers` can go in now.


Of course, if you choose to do this, you'd want to check that it looks correct.

**Comment by raynelfss** [comment]
Hey @jlapeyre this PR is open and ready for review. All I'm waiting for at the moment is approval from @kevinhartman. All the `layers` method have successfully been implemented, and are just waiting to be merged.

**Comment by kevinhartman** [comment]
I haven't made it through in depth yet, but IMO we ought to merge it and address any issues directly in the PR branch since this work is needed to unblock testing.

Thanks for this work @raynelfss !

**Review by kevinhartman — APPROVED** [review]
Please address merge conflicts and then let's get this in.

And additional fixes needed we will apply directly once merged.

**Comment by jlapeyre** [comment]
I think it should be correct. I 
* Merged `oxidize_dag`  2651500701dd. Note this is before Elena's PR and a few other things that have since been commited on top of this.
* Fixed merge conflicts. These were mostly straightforward, where it was clear what to do.
* There were a few calls to method `op_nodes` that needed to be changed to the new name `py_op_nodes`.
* Then the merged branch compiled cleanly modulo the tons of warnings.
* I scanned the failures. There are a bunch, but none appeared to be related to merging this. Nothing related to layers. I may have missed something.
* More than 1000 new tests ran.  Some success and some failure.


I'll look at the new merge conflicts.

**Comment by jlapeyre** [comment]
https://github.com/jlapeyre/qiskit-core/tree/oxidize-dag-layers-merge-base now has no merge conflicts. It is this branch `oxidize-dag-layer` with the latest c1ba59682 `oxidize-dag` merged in.

 This of course does not include commits that may have been made to `main` in the meantime.

Most of the 16 panics are from remaining `todo()!`s. There are a few that are not.

**Comment by raynelfss** [comment]
Oh @jlapeyre I was working on updating this branch to be up-to-date with main. Should I wait for your changes???

**Comment by jlapeyre** [comment]
@raynelfss this looks like great work to me!

**Comment by jlapeyre** [comment]
Yes definitley wait! My branch includes main merged in this afternoon.

**Comment by jlapeyre** [comment]
I was thinking of doing it in two stages. But there is no reason. Let me try to merge main right now.

**Comment by raynelfss** [comment]
Oh okay, I had it already merged (but not pushed), there weren't many issues with this branch, just a couple of wrong calls to `op_nodes` instead of `py_op_nodes`, and other small things. I will keep an eye on how this goes.

**Comment by jlapeyre** [comment]
Sorry about the confusion. You probably made the same changes I did. I merged main, which was one commit/PR, and am now testing. EDIT: I had to throw away a few hours of  work today becaues @mtreinish beat me to it! The "work" was mostly puzzling over code.

**Comment by jlapeyre** [comment]
I merged main just now. When I first merged into this branch a few hours ago 6500 tests ran. Now 5300 tests run. My best guess is that these are new bugs that are panicking and preventing following tests from running.

I think it makes sense to accept this and then fix those bugs in `oxidize-dag`.

**Comment by jlapeyre** [comment]
> couple of wrong calls to `p_nodes` instead of `py_op_nodes`

That's what I saw too.

**Comment by raynelfss** [comment]
@jlapeyre that's interesting, I just ran the tests in my branch and 6654 tests ran.
```bash
======
Totals
======
Ran: 7817 tests in 35.7620 sec.
 - Passed: 6654
 - Skipped: 83
 - Expected Fail: 2
 - Unexpected Success: 0
 - Failed: 1078
Sum of execute time for each test: 150.6633 sec.

==============
Worker Balance
==============
 - Worker 0 (261 tests) => 0:00:03.093905
 - Worker 1 (261 tests) => 0:00:06.680893
 - Worker 2 (262 tests) => 0:00:04.105740
 - Worker 3 (264 tests) => 0:00:08.181179
 - Worker 4 (249 tests) => 0:00:02.660219
 - Worker 5 (274 tests) => 0:00:08.825780
 - Worker 6 (968 tests) => 0:00:12.883057
 - Worker 7 (1097 tests) => 0:00:12.096174
 - Worker 8 (245 tests) => 0:00:02.870246
 - Worker 9 (961 tests) => 0:00:18.592182
 - Worker 10 (949 tests) => 0:00:11.584417
 - Worker 11 (272 tests) => 0:00:04.936310
 - Worker 12 (235 tests) => 0:00:05.150038
 - Worker 13 (993 tests) => 0:00:12.404244
 - Worker 14 (263 tests) => 0:00:05.747054
 - Worker 15 (263 tests) => 0:00:35.135843
```

**Comment by jlapeyre** [comment]
What is the last commit in your branch? Or the last version of `main`?  Does it include 5a6312d7 ?

**Comment by jlapeyre** [comment]
It said 7817 tests ran!

**Comment by jlapeyre** [comment]
I did not realize this PR was essentially done. @raynelfss also merged `oxidize-dag` and introduced fewer bugs than I did. So his commit should be used, not mine.

**Comment by jlapeyre** [comment]
@kevinhartman  we will wait for you to merge this.

**Review by jlapeyre — APPROVED** [review]

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:3293]: Good catch regarding the wrong name in the stub :)

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:3398]: Better to raise an error or at least panic if something weird like the node not being found happens, I'd say!

File: crates/circuit/src/dag_circuit.rs
Comment (kevinhartman) [inline crates/circuit/src/dag_circuit.rs:4556]: I am not keen on using Box unless it is strictly necessary, but we can easily change this later :)

**Review by kevinhartman — APPROVED** [review]
This looks good, I like the way you've refactored the key-based sort into a separate function for clean code reuse. Thanks for the great work, @raynelfss, and thanks for helping to get this merged @jlapeyre.

**Comment by jlapeyre** [comment]
~Addendum: I ran the test suite locally and it shows the same number of tests run as I found on my branch-- 6551. So we introduced the same number of bugs, hopefully zero.  Maybe the 7000 was for something in main not yet merged? Doesn't matter much now.~

The whole story about differing numbers of tests run is due to randomness in the test suite.

 Great work @raynelfss  and @kevinhartman .

**Pull request merged**
